libs/GraphicECS/SFML/Systems/InputManagement.cpp

```diff
@@ -124,9 +124,42 @@ namespace graphicECS::SFML::Systems
                 if (actions.front().first == ActionQueueComponent::BUTTON_CLICK) {
                     clickHandle(world, actions.front().second);
                 }
-                actions.pop();
+                {
+                    auto entity_guard = std::lock_guard(*entityPtr.get());
+                    actions.pop();
+                }
             }
         }
+        shoot(world);
+    }
+
+    void InputManagement::shoot(World &world)
+    {
+        std::vector<std::shared_ptr<Entity>> player = world.joinEntities<Controlable>();
+
+        if (player.empty())
+            return;
+        auto shoot = [&world](std::shared_ptr<Entity> entityPtr) {
+            if (entityPtr->contains<IsShootingComponent>()) {
+                ShootingFrequency &freq = entityPtr.get()->getComponent<ShootingFrequency>();
+                const char hex_char[] = "0123456789ABCDEF";
+                RandomDevice &random = world.getResource<RandomDevice>();
+                std::string uuid(16, '\0');
+
+                if (freq.frequency == duration<double>(0.0)) {
+                    for (auto &c : uuid) {
+                        auto guard = std::lock_guard(random);
```

Comment:
Perhaps you should place this lock just before the for loop, in a different scope. It's a little more optimized.
Or use .lock() and .unlock() before and after the loop